File: binary_search/binary_search.js
function binarySearch(arr, target) {
  let start = 0,
    end = arr.length - 1,
    middle = Math.floor((start + end) / 2);

  while (start <= end) {
    if (target < arr[middle]) end = middle - 1;
    else start = middle + 1;

    // As we update the start and end, have to
    // recalculate new middle
    middle = Math.floor((start + end) / 2);
  }

  if (arr[middle] !== target) return -1;

  return middle;
}

console.log(binarySearch([2, 5, 6, 9, 13, 15, 28], 28));
console.log(binarySearch([28, 15, 13, 9, 6, 5, 2], 2));

File: binary_search/binary_search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums) - 1
        is_ascending = nums[start] < nums[end]

        while start <= end:
            # Calculate the middle of the current range
            mid = start + (end - start) // 2

            if target == nums[mid]:
                return mid

            if is_ascending: # Ascending order
                if nums[mid] > target:
                    end = mid - 1  # The 'target' can be in the first half
                else:  # target > nums[mid]
                    start = mid + 1  # The 'target' can be in the second half
            else:  # Descending order
                if nums[mid] > target:
                    end = mid - 1  # The 'target' can be in the first half
                else:  # target < nums[mid]
                    start = mid + 1  # The 'target' can be in the second half

        return -1  # Element not found

# Time Complexity: Since, we are reducing the search range by half at every
# step, this means that the time complexity of our algorithm will be
# O(logN) where ‘N’ is the total elements in the given array.

# Space Complexity: The algorithm runs in constant space O(1).